### PR TITLE
Implemented Pagination and Backend for the Main Page

### PR DESCRIPTION
**Pull request by Lord-Bubbles**

### CODE REVIEW

**Comment by Lord-Bubbles** [comment]
the look of main page with 5 recipes per page: https://drive.google.com/file/d/1KOgSUm5OWaSSAQV4B6dfBb2yaqKlx7ju/view?usp=sharing

**Comment by Lord-Bubbles** [comment]
#17 #19 #20 #21 #12

File: package.json [hunk 1]
```diff
@@ -1,20 +1,25 @@
 {
   "name": "recipe-app",
   "version": "0.1.0",
+  "proxy": "http://localhost:8080",
```
Comment (chilimalone) [inline package.json:4]: don't mark anything as localhost, because this won't always work when you're running in production

File: src/App.js [hunk 2]
```diff
@@ -1,26 +1,52 @@
...
-import SearchAndFilter from './components/SearchAndFilter';
 import jwt_decode from "jwt-decode";
-import 'bootstrap/dist/css/bootstrap.min.css';
 import {Dropdown, DropdownButton}  from "react-bootstrap";
-import MyRecipes from './components/MyRecipes';
-import MyMealPlan from './components/MyMealPlan';
+import 'bootstrap/dist/css/bootstrap.min.css';
+import Header from "./components/Header";
+
+const loadScript = (src) =>
```
Comment (chilimalone) [inline src/App.js:8]: why are you loading additional scripts on the fly?

File: src/components/Header.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
+import React from "react";
```
Comment (chilimalone) [inline src/components/Header.js:1]: extracting the header into a separate file should really have been a separate pull request

File: src/components/Header.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    this.setState({ ingredients: ingredients });
+  }
+
+  // Call back function to set creator
+  setCreator = (user) => {
+    this.setState({ createdBy: user });
+  }
+
+  // Filters what gets displayed based on the user inputs
+  fetchCurRecipes = async () => {
```
Comment (chilimalone) [inline src/components/Header.js:46]: fetchCurrentRecipes - spell it all out

File: src/components/RecipeView.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    const recipes = this.props.data.map((recipe) => {
+      return <Recipe data={recipe} key={recipe.id}/>
+    })
+    return recipes;
+  }
+
+  render() {
+    return (
+      <div className="container">
+        {this.createRecipes()}
```
Comment (chilimalone) [inline src/components/RecipeView.js:18]: i'd like a better name than createRecipes. maybe "populateRecipes"?

create in context of your application conflates itself with actual creation of recipes, which is confusing when you also have pagination going on.

File: src/components/RecipeView.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import React from "react";
+import Pagination from "./Pagination";
+import Recipe from "./Recipe";
+import "../styles/RecipeView.css";
+
+class RecipeView extends React.Component {
```
Comment (chilimalone) [inline src/components/RecipeView.js:6]: Add a comment at the beginning detailing what view exactly this is - we now have Recipe, RecipeView, etc, which is confusing what view is what.

File: src/components/SearchAndFilter.js [hunk 7]
```diff
@@ -9,68 +9,24 @@ class SearchAndFilter extends React.Component {
     super(props);
     this.state = {
       modalOpen: false,
-      name: "",
-      ingredients: [],
-      minTime: 0,
-      maxTime: 0,
-      createdBy: "",
+      name: this.props.name,
```
Comment (chilimalone) [inline src/components/SearchAndFilter.js:12]: this should've been in a separate pull request.

File: src/database/RecipeEntity.js [hunk 8]
```diff
@@ -1,3 +1,84 @@
...
+  const info = results[1];
+
+  if (info.moreResults !== Datastore.NO_MORE_RESULTS) {
+    const results = await getRecipes(info.endCursor);
+    return [...entities, results[0]];
+  }
+  return entities;
+}
+
+async function getRecipes(data) {
```
Comment (chilimalone) [inline src/database/RecipeEntity.js:22]: nit: data -> filterparams?

Or a comment in the method for what data is supposed to be.

File: src/database/RecipeEntity.js [hunk 9]
```diff
@@ -1,3 +1,84 @@
...
+        time: recipe.time
+      };
+    });
+    data.recipes = JSON.stringify(obj);
+  }
+  let recipes = JSON.parse(data.recipes);
+  return filterRecipes(data, recipes);
+}
+
+function filterRecipes(data, recipes) {
```
Comment (chilimalone) [inline src/database/RecipeEntity.js:41]: ditto comment about parameters

File: src/database/RecipeEntity.js [hunk 10]
```diff
@@ -1,3 +1,84 @@
...
+        name: recipe.name,
+        ingredients: recipe.ingredients,
+        method: recipe.method,
+        createdBy: recipe.createdBy,
+        time: recipe.time
+      };
+    });
+    data.recipes = JSON.stringify(obj);
+  }
+  let recipes = JSON.parse(data.recipes);
```
Comment (chilimalone) [inline src/database/RecipeEntity.js:37]: I don't understand this - you stringify the recipe objects, and then you parse it again? Why not just leave them in data format?

File: src/database/RecipeEntity.js [hunk 11]
```diff
@@ -1,3 +1,84 @@
...
+  return filterRecipes(data, recipes);
```
Comment (chilimalone) [inline src/database/RecipeEntity.js:38]: This implies that you are getting ALL the recipes from your database and then doing filtering one by one through code. This is incredibly inefficient and taxing on memory (what if you've 4 billion recipes? How to keep all of them in memory when array index is going to go out of bounds?)

Instead, you should make sure the results you get from the database is already filtered. Your database query should support things like limit, filter, groupby, start, etc. You can also select specific columns rather than get everything.

e.g. datastore.createQuery("Recipes").limit(10).filter('name', '=', 'Eggfooyoung')

Look into sql queries and capabilities if this doesn't make sense to you. The nodejs api reference for query is at https://cloud.google.com/nodejs/docs/reference/datastore/latest/datastore/query#_google_cloud_datastore_Query_filter_member_1_

File: src/database/UserEntity.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
+const {Datastore} = require("@google-cloud/datastore");
+const datastore = new Datastore();
+
+async function addUsers(data) {
+  const userKey = datastore.key("User");
```
Comment (chilimalone) [inline src/database/UserEntity.js:5]: note - This should've been yet another pull request

File: src/styles/Recipe.css [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+.modal-box {
+  border-style: ridge;
+  outline: black;
+  background-color: transparent;
+}
+
+.modal-time {
+  display: inline;
+  padding-left: 75%;
+}
```
Comment (chilimalone) [inline src/styles/Recipe.css:19]: newline at end of file

File: src/styles/RecipeView.css [hunk 14]
```diff
@@ -0,0 +1,5 @@
+.pagination-bar {
+  width: 100%;
+  display: flex;
+  justify-content: center;
+}
```
Comment (chilimalone) [inline src/styles/RecipeView.css:5]: newline at end of file

File: src/styles/Pagination.scss [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        border-top: 0.12em solid rgba(0, 0, 0, 0.43);
+      }
+
+      &:hover {
+        background-color: transparent;
+        cursor: default;
+      }
+    }
+  }
+}
```
Comment (chilimalone) [inline src/styles/Pagination.scss:69]: newline

**Review by chilimalone — CHANGES_REQUESTED** [review]
There is no reason this pull request should be so big - This could have been broken down into 3 or 4 different pull requests:
1. split out header into separate component
2. Create Recipes component and frontend pagination
3. Server side recipes and pagination support
4. server-side Save and retrieve user

In future, please consider smaller chunks of code in a pull request so you can help your reviewer focus, but also allows you to make progress / test things along the way instead of one giant thing

In addition, smaller focused code chunks means you only roll back a small chunk if something goes wrong (e.g. if pagination ended up causing 500 server errors). Submitting code in a massive chunk like this just means *everything* gets rolled back if only a small part goes wrong.

File: package.json (hunk 1 above)
Comment (Lord-Bubbles) [reply package.json:4]: done

File: src/App.js (hunk 2 above)
Comment (Lord-Bubbles) [reply src/App.js:8]: removed
